Sources/ClientRuntime/Networking/Streaming/Stream.swift

```diff
@@ -0,0 +1,63 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import AwsCommonRuntimeKit
+
+/// Protocol that provides reading data from a stream
+public protocol ReadableStream: AnyObject {
+    /// Returns the current position in the stream
+    var position: Data.Index { get }
+
+    /// Returns the length of the stream, if known
+    var length: Int? { get }
+
+    /// Returns true if the stream is empty, false otherwise
+    var isEmpty: Bool { get }
+
+    /// Returns true if the stream is seekable, false otherwise
+    var isSeekable: Bool { get }
+
+    /// Reads up to `count` bytes from the stream
+    /// - Parameter count: maximum number of bytes to read
+    /// - Returns: Data read from the stream, or nil if the stream is closed and no data is available
+    func read(upToCount count: Int) throws -> Data?
+
+    /// Reads all remaining bytes from the stream
+    /// - Returns: Data read from the stream, or nil if the stream is closed and no data is available
+    func readToEnd() throws -> Data?
+
+    /// Seeks to the specified offset within the stream
+    /// - Parameter offset: offset to seek to
+    func seek(toOffset offset: Int) throws
+}
+
+/// Protocol that provides writing data to a stream
+public protocol WriteableStream: AnyObject {
+    ///  Writes the contents of `data` to the stream
+    /// - Parameter data: data to write
+    func write(contentsOf data: Data) throws
+
+    /// Closes the stream
+    func close() throws
+}
+
+/// Protocol that provides reading and writing data to a stream
+public protocol Stream: ReadableStream, WriteableStream {
+}
+
+public enum StreamError: Error {
+    case invalidOffset(String)
+    case notSupported(String)
+}
+
+extension Stream {
+    func isEqual(to other: Stream) throws -> Bool {
+        let selfData = try readToEnd()
+        let otherData = try other.readToEnd()
+        return selfData == otherData
```

Comment:
Rather than make this throwing, should we swallow the error and assume equality is false in the case where reading either of the streams fails? That simplifies things and seems to be a reasonable tradeoff in this scenario.

Reply:
that seems wrong to me, consumers should care about it then here. Having to bubble the error up forces us to decide better whether do we want to force/swallow the error.

Reply:
Hmm, i don't particularly like that we will crash in the case where this throws, since we are force trying in the places we use this. We shouldn't crash a customers app because reading a stream failed.